### PR TITLE
Ensure that Marshal raises only SwaggerMappingError

### PR DESCRIPTION
**Pull request by macisamuele**
This PR makes sure that while formatting/marshaling a primitive type only ``SwaggerMappingError`` could be raised in case of error (as declared by ``marshal_primitive``'s docstring).

Without the code change introduced by this PR ``test_marshal_raises_SwaggerMappingError_if_SwaggerFormat_fails_during_to_wire`` will fail with the following stacktrace
```
F
tests/marshal/marshal_schema_object_test.py:71 (test_marshal_raises_SwaggerMappingError_if_SwaggerFormat_fails_during_to_wire)
minimal_swagger_dict = {'definitions': {'test': {'properties': {'date': {'format': 'date', 'type': 'string'}}, 'required': ['date'], 'type': 'object'}}, 'info': {'title': 'Test', 'version': '1.0'}, 'paths': {}, 'swagger': '2.0'}

    def test_marshal_raises_SwaggerMappingError_if_SwaggerFormat_fails_during_to_wire(minimal_swagger_dict):
        minimal_swagger_dict['definitions']['test'] = {
            'properties': {
                'date': {
                    'type': 'string',
                    'format': 'date',
                },
            },
            'required': [
                'date',
            ],
            'type': 'object'
        }
        swagger_spec = Spec(minimal_swagger_dict)
        with pytest.raises(SwaggerMappingError):
            marshal_schema_object(
                swagger_spec=swagger_spec,
                schema_object_spec=minimal_swagger_dict['definitions']['test'],
>               value={'date': datetime.date.today().isoformat()},
            )

marshal_schema_object_test.py:90: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../bravado_core/marshal.py:54: in marshal_schema_object
    return marshal_object(swagger_spec, schema_object_spec, value)
../../bravado_core/marshal.py:147: in marshal_object
    result[k] = marshal_schema_object(swagger_spec, prop_spec, v)
../../bravado_core/marshal.py:37: in marshal_schema_object
    return marshal_primitive(swagger_spec, schema_object_spec, value)
../../bravado_core/marshal.py:84: in marshal_primitive
    value = formatter.to_wire(swagger_spec, primitive_spec, value)
../../bravado_core/formatter.py:67: in to_wire
    return formatter.to_wire(value)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

d = '2017-09-03'

>   to_wire=lambda d: d.isoformat(),
    to_python=lambda d: dateutil.parser.parse(d).date(),
    validate=NO_OP,  # jsonschema validates date
    description='Converts [wire]string:date <=> python datetime.date'),
        # Python has no double. float is C's double in CPython
        'double': SwaggerFormat(
    format='double',
    to_wire=lambda d: d if isinstance(d, float) else float(d),
    to_python=lambda d: d if isinstance(d, float) else float(d),
    validate=NO_OP,  # jsonschema validates number
    description='Converts [wire]number:double <=> python float'),
        'date-time': SwaggerFormat(
    format='date-time',
    to_wire=lambda dt: (dt if dt.tzinfo else pytz.utc.localize(dt)).isoformat(),
    to_python=lambda dt: dateutil.parser.parse(dt),
    validate=NO_OP,  # jsonschema validates date-time
    description=(
        'Converts string:date-time <=> python datetime.datetime')),
        'float': SwaggerFormat(
    format='float',
    to_wire=lambda f: f if isinstance(f, float) else float(f),
    to_python=lambda f: f if isinstance(f, float) else float(f),
    validate=NO_OP,  # jsonschema validates number
    description='Converts [wire]number:float <=> python float'),
        'int32': SwaggerFormat(
    format='int32',
    to_wire=lambda i: i if isinstance(i, int) else int(i),
    to_python=lambda i: i if isinstance(i, int) else int(i),
    validate=NO_OP,  # jsonschema validates integer
    description='Converts [wire]integer:int32 <=> python int'),
        'int64': SwaggerFormat(
    format='int64',
    to_wire=lambda i: i if isinstance(i, long) else long(i),
    to_python=lambda i: i if isinstance(i, long) else long(i),
    validate=NO_OP,  # jsonschema validates integer
    description='Converts [wire]integer:int64 <=> python long'),
    }
E   AttributeError: 'str' object has no attribute 'isoformat'

../../bravado_core/formatter.py:119: AttributeError
```

### CODE REVIEW

File: bravado_core/formatter.py [hunk 1]
```diff
@@ -11,13 +11,40 @@
...
 from bravado_core import schema
+from bravado_core.exception import SwaggerMappingError
 
 if six.PY3:
     long = int
 
 
 def NO_OP(x):
-    return None
+    return x
```
Comment (sjaensch) [inline bravado_core/formatter.py]: Heh, that was not a no_op. :)

File: bravado_core/formatter.py [hunk 2]
```diff
@@ -29,12 +56,20 @@ def to_wire(swagger_spec, primitive_spec, value):
...
         return value
     format_name = schema.get_format(swagger_spec, primitive_spec)
-    formatter = swagger_spec.get_format(format_name)
-    return formatter.to_wire(value) if formatter else value
+    formatter = swagger_spec.get_format(format_name) or _NO_OP_SWAGGER_FORMAT
+
+    try:
+        return formatter.to_wire(value)
+    except Exception as e:
+        raise SwaggerMappingError('Error while marshaling value={} to type={}{}.'.format(
```
Comment (sjaensch) [inline bravado_core/formatter.py]: marshalling

File: tests/marshal/marshal_schema_object_test.py [hunk 3]
```diff
@@ -68,6 +69,28 @@ def test_ref(minimal_swagger_dict):
...
+            'date',
+        ],
+        'type': 'object'
+    }
+    swagger_spec = Spec(minimal_swagger_dict)
+    with pytest.raises(SwaggerMappingError):
+        marshal_schema_object(
+            swagger_spec=swagger_spec,
+            schema_object_spec=minimal_swagger_dict['definitions']['test'],
+            value={'date': datetime.date.today().isoformat()},
```
Comment (sjaensch) [inline tests/marshal/marshal_schema_object_test.py]: Two things:
1. could you assert on the error message as well? We've had bugs there before, and it's very important we get it right.
2. Why is this failing? It's not immediately obvious to me.

File: bravado_core/formatter.py [hunk 4]
```diff
@@ -29,12 +56,20 @@ def to_wire(swagger_spec, primitive_spec, value):
...
     :type value: int, long, float, boolean, string, unicode, object, etc
     :rtype: int, long, float, boolean, string, unicode, etc
+    :raises: SwaggerMappingError when format.to_wire raises an exception
     """
     if value is None or not schema.has_format(swagger_spec, primitive_spec):
         return value
     format_name = schema.get_format(swagger_spec, primitive_spec)
-    formatter = swagger_spec.get_format(format_name)
-    return formatter.to_wire(value) if formatter else value
+    formatter = swagger_spec.get_format(format_name) or _NO_OP_SWAGGER_FORMAT
```
Comment (sjaensch) [inline bravado_core/formatter.py]: Why is this necessary? It looks like we're adding one function call for the marshalling of each primitive value (as I assume not having a format is the common case). I'm not sure if this will be noticeable overhead, but can't we simply keep the behavior of short-circuiting and returning value?

File: bravado_core/formatter.py (hunk 1 above)
Comment (macisamuele) [reply bravado_core/formatter.py]: Just to be sure, are you asking me to revert this change?

Comment (sjaensch) [reply bravado_core/formatter.py]: Nope, just commenting on the fact that you're fixing something along the way. :)
